### ISSUE TITLE
Framework: Changing between loaded Devdocs pages lags

### ISSUE THREAD

**Issue by tofumatt** [Automattic/wp-calypso#22756]
#### Steps to reproduce
1. Starting at URL: http://calypso.localhost:3000/devdocs/design
2. Navigate to the blocks page (http://calypso.localhost:3000/devdocs/blocks)
3. Click the UI Components link to load that page again
4. Wait a few seconds to a screen with no loading indicator while hundreds of components are loaded

#### What I expected
Components are lazy rendered or an interstitial loading screen is displayed.

#### What happened instead
The page loads but no screen/indicator is given and it seems like nothing is happening.

This is more apparent after #20690 because the first time the page loads we show a placeholder page as the components are asynchronously loaded.

#### Screenshot / Video
![2018-02-23 14 48 31](https://user-images.githubusercontent.com/90871/36599938-d053dfb0-18a8-11e8-84e7-fe1aa489e6dc.gif)

**Comment by tofumatt** [comment]
I took a look at this and the main issue seems to be the number of components loaded on the page. The render ends up mounting lots of children components and it gets pretty slow.

My first instinct was to try a virtualised list that loaded only the components in the viewport. This actually did end up decreasing the initial load by up to 2 seconds, but resulted in:

1. a more complex code path
2. slower scrolling performance as element size and which ones to display had to be calculated
3. a worse scrolling experience as all pages elements weren't rendered
4. the lack of ability to use the browser's find functionality as all page elements weren't rendered

That said, there might be a way to speed things up inside the render method and in some other ways. So far virtualised lists don't seem to do the trick though.

**Comment by tofumatt** [comment]
Some of the stuff in https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html looks promising here too.

**Comment by tofumatt** [comment]
I talked to @nb about this but it seems the main issue is around the (very large) number of components mounted inside these pages (`/devdocs/blocks` and `/devdocs/design`). Pagination through the components wouldn't be a good UX so continuing to load all of the examples on the page is what I aimed to keep.

First off I tried using something like [`react-virtualized`](https://github.com/bvaughn/react-virtualized) to load only the content on screen but because the components are of varying heights that lead to only a ~10-20% reduction in mounting time and a worse scrolling UX. It also meant the entire page would never be mounted at once so in-browser "find" UX was ruined.

Using [`react-lazily-render`](https://github.com/jameslnewell/react-lazily-render) turned out to work much better, and using it means we can load the first X number of components immediately (basically enough to fill the viewport) so the component mounts in a fraction of the time it did in the past (for reference, it mounts faster using React development environment with the change than it currently does in production with the old "mount everything at once" approach). We can then load the rest of the components either: onScroll or after a few seconds with a timeout. The underlying code used is a bit gnarly but the API of the Devdocs `Collection` remains the same with the heavy-lifting/lazy-loading happening transparently, so this approach will scale if more pages are added.

The end result is changing between these large pages feels faster. I'll have a PR shortly, but wanted to log my findings here for now 😄
